### PR TITLE
Pre-Order Validation

### PR DESCRIPTION
**Pull request by bh2smith**
In Preparation for the new Fee and Quote endpoint (cf #1155) we shift some of the order validation into its own file (to be shared among both `Orderbook.add_order` and the `QuoteRequest` endpoint). More specifically, we move all order validations pertaining to the buy/sell tokens, order validity, balance sources and destinations, banned users, and transfer eth to contract into their own `PreOrderValidator`. Any order validations regarding balance or fee sufficiency and signature validity remain in the `Orderbook.add_order` method.

### Test Plan
This is a refactor although I have added some tests to the new method `validate_partial_order`

### CODE REVIEW

File: orderbook/src/api/validation.rs [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+#[derive(Default)]
+pub struct PreOrderData {
+    owner: H160,
+    sell_token: H160,
+    buy_token: H160,
+    receiver: H160,
+    valid_to: u32,
+    buy_token_balance: BuyTokenDestination,
+    sell_token_balance: SellTokenSource,
```
Comment (vkgnosis) [inline orderbook/src/api/validation.rs:75]: Probably want all of these `pub` too.

**Review by vkgnosis — APPROVED** [review]

File: orderbook/src/api/validation.rs (hunk 1 above)
Comment (bh2smith) [reply orderbook/src/api/validation.rs:75]: Yea good call, I made these all public in the next PR, but I can do it here in favour of less changes in the next one.

File: orderbook/src/api/validation.rs [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+                sell_token: BUY_ETH_ADDRESS,
+                buy_token: native_token.address(),
+                ..Default::default()
+            },
+            &native_token,
+        ));
+    }
+
+    #[tokio::test]
+    async fn validate_partial_order_err() {
```
Comment (nlordell) [inline orderbook/src/api/validation.rs:205]: One underrated feature of this refactor is the ability to test order validation better.

**Review by nlordell — APPROVED** [review]
Love it.

File: orderbook/src/api/validation.rs (hunk 2 above)
Comment (bh2smith) [reply orderbook/src/api/validation.rs:205]: Perhaps in a follow up PR, I can make a similar method for the other parts of order validation.

**Pull request merged**